x-pack/plugins/enterprise_search/server/lib/ml_inference_pipeline/get_inference_pipelines.ts

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IngestPipeline, IngestProcessorContainer } from '@elastic/elasticsearch/lib/api/types';
+import { ElasticsearchClient } from '@kbn/core/server';
+import { MlTrainedModels } from '@kbn/ml-plugin/server';
+
+import { getMlModelConfigsForModelIds } from '../indices/fetch_ml_inference_pipeline_processors';
+
+/**
+ * Gets all ML inference pipelines. Redacts trained model IDs in those pipelines which reference
+ * a model inaccessible in the current Kibana space.
+ * @param esClient the Elasticsearch Client to use to fetch the errors.
+ * @param trainedModelsProvider ML trained models provider.
+ */
+export const getMlInferencePipelines = async (
+  esClient: ElasticsearchClient,
+  trainedModelsProvider: MlTrainedModels | undefined
+): Promise<Record<string, IngestPipeline>> => {
+  if (!trainedModelsProvider) {
+    return Promise.reject(new Error('Machine Learning is not enabled'));
+  }
+
+  // Fetch all ML inference pipelines
+  const fetchedInferencePipelines = await esClient.ingest.getPipeline({
+    id: 'ml-inference-*',
+  });
+
+  // Fetch all trained models from the current Kibana space, and identify those that have their
+  // model_id redacted
+  const trainedModels = await getMlModelConfigsForModelIds(esClient, trainedModelsProvider);
```

Comment:
I'm not sure we **should** re-use `getMlModelConfigsForModelIds` here since all we want is a list of models that need to be redacted. My reason for saying that is `getMlModelConfigsForModelIds` fetches all the models and modelStats which we don't need here. It's making 3 api calls when I think we only need to get the models in the space and compare that against the `fetchedInferencePipelines` and redact models that are not in the space.

Reply:
I was going to take a closer look at `getMlModelConfigsForModelIds()` after the refactoring I'm proposing - and as part of that, making it a shared function and breaking it up into smaller sub-functions -, but your suggestion looks good too and simple. We just need to reverse the logic when identifying redacted model IDs (whitelist instead of blacklist).

Reply:
This is intended to be used by the Add Inference Pipeline modal that currently gets the trained models from the ml api and then we need to query them again here. It makes me wonder if we should have a single endpoint for the data the modal needs. Or stick with the current plan to have it make multiple GET requests for specific data, even if those requests end up making the same queries under-the-hood. 😕

Reply:
What's our team's best practice for this - each modal typically gets its data in a single request? (Versus each widget in a modal.)

I can see a potential drawback in that those endpoints will be very specfic to the modal's use case and not reusable in any other context. If we used GraphQL, that would be a different story, but with REST we need to find the right balance.

My preference is to finish this API as we discussed (`GET /internal/enterprise_search/pipelines/ml_inference` will get ML inference pipelines as it suggests), and then see if it makes sense to refactor into / add a different endpoint for the modal. How does that sound?

Reply:
> What's our team's best practice for this 

I don't know

> My preference is to finish this API as we discussed

I agree.